### ISSUE TITLE
Barely any difference with network_alpha = 1 even on high lr

### ISSUE THREAD

**Issue by anon-1337** [kohya-ss/sd-scripts#98]
![xy_grid-0116-NAI](https://user-images.githubusercontent.com/122653835/213949412-9542e0f3-afcd-4c8c-8f68-1bf9abefbf71.png)
![xy_grid-0117-NAI](https://user-images.githubusercontent.com/122653835/213949417-014c5e6a-787b-48fa-ab9a-03b9f4326533.png)
Image difference between one of the gens (0.4.0 network with alpha = 1 and image without network)
![gimp-2 10_7SG0N1c7OH](https://user-images.githubusercontent.com/122653835/213950527-38381a84-7b6c-4850-b821-674ff8e7a7ec.png)

kek | Monika v1 (pre0.4.0) | Monika v1 fix | Monika v1 fix lr1e-3
------ | ------ | ------ | ------
learning_rate | 2e-4 | 2e-4 | 1e-3
text_encoder_lr | 5e-5 | 5e-5 | -
unet_lr | 3e-4 | 3e-4 | -
network_dim | 128 | 128 | 128
network_alpha | - (128 I guess) | 1 | 1
commit | 22ee0ac467ffc914fe174b7006308d6cbf7a6f63 | 93df55d5976c9b9c50c81a62174f1c4e5e5d001d | 93df55d5976c9b9c50c81a62174f1c4e5e5d001d

All other settings are the same: seed = 1337, shuffle_caption disabled. Same dataset, model, VAE.
Using sd-additional-networks commit [15ee6994f2723970eda300f1bb0dc917c44636f3](https://github.com/kohya-ss/sd-webui-additional-networks/commit/15ee6994f2723970eda300f1bb0dc917c44636f3)

Layers weights difference:
[Monika v1 vs Monika v1 fix](https://diffonline.net/n9Jukz9JLe)
[Monika v1 vs Monika v1 fix lr1-e3](https://diffonline.net/sUHT8LuTwe)

I have another question though, is ```learning_rate``` like a different thing from ```text_encoder_lr``` and ```unet_lr```? Or is it just a "master knob" for both learning rates if they are not specified? I'm noob.

Thank you.

**Comment by mgz-dev** [comment]
For learning rate, if you specify `text_encoder_lr` and `unet_lr` then `learning_rate` will not be used (the ones specified for each respective part of the model will be used instead).

As an explanation, we are essentially training on two deep learning model architectures (part of a unet model and part of a clip model/textencoder). We may want to update weights (think of this as the knowledge captured by the lora model) differently for each part, thus the need to specify different learning rates.

The learning rate represents how strongly we want to react in response to a gradient loss observed on the training data at each step (the higher the learning rate, the bigger moves we make at each training step). You can think of loss in simple terms as a representation of how close your model prediction is to a true label.

Your understanding of alpha seems on track from my personal understanding. If you look at the `lora.py` file under the `apply_to()` function, you can see that the lora is *added* to the original forward function (the original output of your stable diffusion model at a given layer) but now (with the 0.4.0 changes) modified by a scalar value named `scale` which is `alpha/lora_dim`. 

This means compared to old lora models, the same output values at a given layer effect the combined output less (by a factor of `scale`).  The magnitude of the outputs from the lora net will need to be "larger" to impact the network the same amount as before (meaning the weights within the lora likely will also be larger in magnitude). This is likely why Kohya stated with `alpha=1` we will probably need higher learning rates.

Hopefully this gives some insight on what is actually happening when you're training!

This all said I haven't had a chance yet to do the test you did above so there may be something else awry. Did you also observe no difference when you used lr1e-3?

**Comment by anon-1337** [comment]
@mgz-dev thank you for detailed explanation! Now I understand the math behind a bit.

> Did you also observe no difference when you used lr1e-3? I'll give the updated code a try today.

Yes, image 2 in the first post.

I tried to "multiply magnitudes of the outputs" by increasing both `unet_lr` and `text_encoder_lr` by a factor of 128 to make up of it:
`unet_lr = 3e-4` became `unet_lr = 0.0384`
`text_encoder_lr = 5e-5` became `text_encoder_lr = 0.0064`

But it didn't work. Here are my loss/average and lr/unet graphs. All step losses after step 52 are NaN:

![изображение](https://user-images.githubusercontent.com/122653835/214008871-483ce6b9-70b4-45f6-9391-979d0aecd61a.png)
(some weird behavior on both lr graphs)


<details><summary>train_network.py output</summary>
<p>

![изображение](https://user-images.githubusercontent.com/122653835/214009666-279290a0-1c01-4821-9084-0661b47b0a2b.png)

</p>
</details>

Even sd-additional-networks extension refuses to work:

<details><summary>AUTOMATIC1111-webui output</summary>
<p>

![изображение](https://user-images.githubusercontent.com/122653835/214009979-0fbe4052-d84d-4ad6-84df-eb07adcee51c.png)

</p>
</details>

[LoRA network weights](https://rentry.org/6qu34)

**Comment by mgz-dev** [comment]
It looks like you are getting NaN as loss values on the chart? The learning rate may be too high (do not increase it linearly with the dimension of the lora).

Here is an example tensorboard plot from my previous run (smoothing is on):

![image](https://user-images.githubusercontent.com/49577754/214093634-5bb558d0-03a2-41eb-9fa5-692967cfc666.png)

```   
    "batch_size": 3,
    "unet_lr": 0.005,
    "text_encoder_lr": 0.001,
    "network_alpha": 1,
    "mixed_precision": "fp16",
    "save_precision": "fp16"
```

Again - these learning rates I used were too high for my dataset (a "good" LR will vary depending on dataset, base model, batch size, etc) but still gave me an overfit lora.

Also I'd recommend checking to see if you are using the versions of xformers and pytorch specified in the requirements.txt file since you're having problems. I had issues with newer versions of those two packages and had to roll the versions back (though maybe they've been fixed with the latest release).

**Comment by anon-1337** [comment]
Jesus f-g christ all of the problems were because of A1111's webui was throwing me errors. And I thought there was something wrong with training process. Just git pulled webui and everything works fine.
But I need to figure out new learning rates with alpha = 1 to replicate old "broken" networks anyway.
Thank you for help.
